api/src/migrations/api/1619711501000-DropNotNullsOnTimeUserMetadataColumns.ts

```diff
@@ -0,0 +1,50 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+import { v4 } from 'uuid';
+
+export class DropNotNullsOnTimeUserMetadataColumns1619711501000
+  implements MigrationInterface {
+  private tablesToAlter = ['organizations', 'projects', 'scenarios'];
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await Promise.all(
+      this.tablesToAlter.map(async (table) => {
+        await queryRunner.query(`
+ALTER TABLE ${table}
+  ALTER COLUMN created_at DROP NOT NULL,
+  ALTER COLUMN last_modified_at DROP NOT NULL,
+  ALTER COLUMN created_by DROP NOT NULL;
+`);
+      }),
+    );
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    /**
+     * We first create a dummy user to whom to associate data with no owner so
+     * that the following ALTER TABLE can succeed. This may in fact never be
+     * needed, and it is an arbitrary association of "orphaned" data, but just
+     * in case...
+     */
+    const dummyUserEmailAddress = `${v4()}@example.com`;
+    await queryRunner.query(`
+INSERT INTO users
+  (email, password_hash, display_name, is_active)
+VALUES
+  (${dummyUserEmailAddress}, crypt('${v4()}', gen_salt('bf')), 'Default owner for data imported via ETL jobs', false);
+`);
+    await Promise.all(
+      this.tablesToAlter.map(async (table) => {
+        await queryRunner.query(`
+UPDATE ${table}
+  SET created_by = (SELECT id FROM users WHERE email = '${dummyUserEmailAddress}')
```

Comment:
Shouldn't it set `created_at` and `last_modified_at` as well? (might be default now() as well).

Reply:
yep, the down side of this migration doesn't make fully sense in any case as it makes up data (owner and timestamps) that we cannot have if we haven't been gathering them - timestamps will default to `now()`. I don't envisage this side of the migration to have more than a formal value - if needing to roll back, we won't have SQL errors, but owner/timestamps data may be arbitrary.

Reply:
all good, I wasn't sure if we have `default` set 👍